src/VisualStudio/CodeLens/ReferenceCodeLensProvider.cs

```diff
@@ -0,0 +1,280 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeLens;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Editor.Wpf;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.ServiceHub.Client;
+using Microsoft.VisualStudio.Core.Imaging;
+using Microsoft.VisualStudio.Language.CodeLens.Remoting;
+using Microsoft.VisualStudio.Language.Intellisense;
+using Microsoft.VisualStudio.LanguageServices.CodeLens;
+using Microsoft.VisualStudio.LanguageServices.Remote;
+using Microsoft.VisualStudio.Threading;
+using Microsoft.VisualStudio.Utilities;
+using Roslyn.Utilities;
+using StreamJsonRpc;
+
+namespace Microsoft.VisualStudio.CodeAnalysis.CodeLens
+{
+    [Export(typeof(IAsyncCodeLensDataPointProvider))]
+    [Name(Id)]
+    [ContentType(ContentTypeNames.CSharpContentType)]
+    [ContentType(ContentTypeNames.VisualBasicContentType)]
+    [LocalizedName(typeof(CodeLensVSResources), "CSharp_VisualBasic_References")]
+    [Priority(200)]
+    [OptionUserModifiable(userModifiable: false)]
+    [DetailsTemplateName("references")]
+    internal class ReferenceCodeLensProvider : IAsyncCodeLensDataPointProvider
+    {
+        // TODO: do we need to localize this?
+        private const string Id = "CSVBReferences";
+
+        // these string are never exposed to users but internally used to identify 
+        // each provider/servicehub connections and etc
+        private const string HubClientId = "ManagedLanguage.IDE.CodeLensOOP";
+        private const string RoslynCodeAnalysis = "roslynCodeAnalysis";
+
+        private readonly HubClient _client;
+
+        // use field rather than import constructor to break circular MEF dependency issue
```

Comment:
😕 Can you explain this in more detail? Why can we not use an importing constructor?

```csharp
[ImportingConstructor]
[Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
public ReferenceCodeLensProvider(Lazy<ICodeLensCallbackService> codeLensCallbackService)
{
}
```

Reply:
there is a reason who want it to be importing constructor? or you just want me to add [Obsolete] ?

and yes. using lazy could solve the circular MEF dependency issue, didn't try. but what is the reason you want to use importing constructor?

Reply:
done